src/index.md

```diff
@@ -58,6 +59,86 @@ const end = view(Inputs.date({label: "End", value: getDateXDaysAgo(1) }));
   }</div>
 </div>
 
+<div class="divider"></div>
+
+<h4>Spark Retrieval Result Codes</h4>
+<body>This section shows the Spark Retrieval Result Codes breakdown. You can adjust the date range. Records exist for the last 90 days.</body>
+
+```js
+const combine = (obj, target, keys) => {
+  obj[target] = keys.reduce((acc, key) => acc + (obj[key] || 0), 0)
+  for (const key of keys) {
+    delete obj[key]
+  }
+}
+const clone = obj => JSON.parse(JSON.stringify(obj))
+const tidy = clone(SparkRetrievalResultCodes).flatMap(({ day, rates }) => {
+  for (const [key, value] of Object.entries(rates)) {
+    rates[key] = Number(value)
+  }
+
+  combine(rates, 'HTTP_5xx', [
+    'LASSIE_502',
+    'HTTP_500',
+    'HTTP_502',
+    'ERROR_503',
+    'ERROR_522',
+    'ERROR_520',
+    'ERROR_500',
+    'BAD_GATEWAY',
+    'GATEWAY_TIMEOUT'
+  ])
+  combine(rates, 'HTTP_4xx', [
+    'ERROR_429',
+    'ERROR_404',
+    'ERROR_403',
+    'ERROR_408'
+  ])
+  combine(rates, 'IPNI_NOT_ADVERTISED', [
+    'IPNI_ERROR_404',
+    'IPNI_NO_VALID_ADVERTISEMENT',
+  ])
+  combine(rates, 'IPNI_4xx', [
+    'IPNI_ERROR_400',
+    'UNSUPPORTED_MULTIADDR_FORMAT'
+  ])
+  combine(rates, 'IPNI_5xx', [
+    'IPNI_ERROR_500',
+    'IPNI_ERROR_502',
+    'IPNI_ERROR_503',
+    'IPNI_ERROR_504',
+    'IPNI_ERROR_FETCH'
+  ])
+  combine(rates, 'CAR_ERROR', [
+    'CANNOT_PARSE_CAR_FILE',
+    'CAR_TOO_LARGE'
+  ])
+
+  return Object.entries(rates).map(([code, rate]) => ({ day: new Date(day), code, rate }))
+})
```

Comment:
Have you considered moving this logic to `src/data/spark-retrieval-result-codes.json.js`? We can write tests for it that way. If the FileAttachment requests are cached, then we will cache the processed data.

Reply:
On the second thought, I think it will be even better to move this to spark-stats, so that everybody else consuming Spark Data API gets the same cleaned-up stats.

Then, it also allows us to add a visualisation to our Grafana dashboards, e.g. the Spark Internal Dashboard.

Reply:
> Have you considered moving this logic to src/data/spark-retrieval-result-codes.json.js? We can write tests for it that way. If the FileAttachment requests are cached, then we will cache the processed data.

This is a pattern in Observable: You have a data source (our JSON API), and then you have a `const tidy = ...` in the notebook, that has the data reshaped as you need it for the visualization. If you look at Observable example notebooks, they all follow this pattern.

> On the second thought, I think it will be even better to move this to spark-stats, so that everybody else consuming Spark Data API gets the same cleaned-up stats.
>
> Then, it also allows us to add a visualisation to our Grafana dashboards, e.g. the Spark Internal Dashboard.

I'm proposing to do this after we have shipped this feature, and gotten feedback at LabWeek, ie when the data model is stable. The advantage of doing it in this repo is that you can iterate on the data and the representation together, and don't need to keep two repositories in sync. The API gives the raw data, the visualization layer can decide how to group. Eg in Grafana you can have a vertical legend, where you don't really need to group the codes. I have done this here primarily because Observable legends are horizontal.